### PR TITLE
[vcpkg] Change scripts/tls12-download.exe mode from 100644 to 100755

### LINKED ISSUE #16211: bootstrap-vcpkg.bat fails when vcpkg repository cloned, if git obeys executable permissions of tls12-download.exe
**Describe the bug**
I just tried cloning the official `vcpkg` repository using Cygwin's `git`, and then tried `bootstrap-vcpkg.bat` from Windows Command Prompt.  These steps had previously worked fine during December 2020,

There seems to be something wrong with one of the executable permissions that had recently been merged into `vcpkg`.

**Environment**
- Windows 10 20H2 19042.804
- Compiler: Visual Studio Community 2019 16.8.5

**To Reproduce**
1. Launch Cygwin Terminal
2. (Cygwin Terminal) `git clone git@github.com:microsoft/vcpkg.git`
3. Launch Windows Command Prompt
4. (Command Prompt) `bootstrap-vcpkg.bat`

**Expected behavior**
4. Bootstrap should not encounter any Access denied errors.

**Failure logs** 
```
C:\src\vcpkg>bootstrap-vcpkg.bat
Program 'tls12-download.exe' failed to run: Access is deniedAt C:\src\vcpkg\scripts\bootstrap.ps1:48 char:1
+ & "$scriptsDir/tls12-download.exe" github.com "/microsoft/vcpkg-tool/ ...
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~.
At C:\src\vcpkg\scripts\bootstrap.ps1:48 char:1
+ & "$scriptsDir/tls12-download.exe" github.com "/microsoft/vcpkg-tool/ ...
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : ResourceUnavailable: (:) [], ApplicationFailedException
    + FullyQualifiedErrorId : NativeCommandFailed


The variable '$LASTEXITCODE' cannot be retrieved because it has not been set.
At C:\src\vcpkg\scripts\bootstrap.ps1:51 char:5
+ if ($LASTEXITCODE -ne 0)
+     ~~~~~~~~~~~~~
    + CategoryInfo          : InvalidOperation: (LASTEXITCODE:String) [], RuntimeException
    + FullyQualifiedErrorId : VariableIsUndefined

Telemetry
---------
vcpkg collects usage data in order to help us improve your experience.
The data collected by Microsoft is anonymous.
You can opt-out of telemetry by re-running the bootstrap-vcpkg script with -disableMetrics,
passing --disable-metrics to vcpkg on the command line,
or by setting the VCPKG_DISABLE_METRICS environment variable.

Read more about vcpkg telemetry at docs/about/privacy.md
```

**Additional context**

Doing a little detective work:

```
$ ls -Al scripts/tls12-download.exe
-rw-r--r-- 1 Jet Jet 18312 Feb 13 18:10 scripts/tls12-download.exe
```
It looks like the executable was cloned and assigned the permissions that had been merged into git, at the time the executable was checked in: microsoft/vcpkg@aa60b7efa56a83ead743718941d8b320ef4a05af

I believe you can overcome this by just setting the mode to 0755, then merging it.  Thiis is the default mode windows assigns executables, if you had just downloaded the executable from the internet:

```
$ chmod 0755 scripts/tls12-download.exe
$ ls -Al scripts/tls12-download.exe
-rwxr-xr-x 1 Jet Jet 18312 Feb 13 18:10 scripts/tls12-download.exe
```

Then from Windows Command-Prompt:
```
C:\src\vcpkg>bootstrap-vcpkg.bat
Downloading https://github.com/microsoft/vcpkg-tool/releases/download/2021-01-13-768d8f95c9e752603d2c5901c7a7c7fbdb08af35/vcpkg.exe -> C:\src\vcpkg/vcpkg.exe
Done.

Telemetry
---------
vcpkg collects usage data in order to help us improve your experience.
The data collected by Microsoft is anonymous.
You can opt-out of telemetry by re-running the bootstrap-vcpkg script with -disableMetrics,
passing --disable-metrics to vcpkg on the command line,
or by setting the VCPKG_DISABLE_METRICS environment variable.

Read more about vcpkg telemetry at docs/about/privacy.md
```

I'll try and open a PR to correct the mode.  This is only .exe in the repository so far, and the problem doesn't seem to affect .bat files.

Otherwise everything else cloned from within Cygwin works great!

### PR DESCRIPTION
**Pull request by yuri-sevatz**
**Describe the pull request**

This PR change the file mode checked into git on `scripts/tls12-download.exe`, from 100644 to 100755.

Correcting the permission merged into git allows `bootstrap-vcpkg.bat` to function without hitting Access denied errors.

Not all git clients are hitting this, but Cygwin's seems to be a little more strict about maintaining an equivalent representation for what was merged, which causes the file to be non-executable at a filesystem level.

Fixes  #16211

### CODE REVIEW

**Comment by yuri-sevatz** [comment]
+ @BillyONeal  (Related to: aa60b7efa56a83ead743718941d8b320ef4a05af)

**Comment by StarGate-One** [comment]
@yuri-sevatz /cc @BillyONeal @ras0219-msft 
- Should one run the `./bootstrap-vcpkg.sh` with Cygwin, just like the rest of the *nix environments MSYS2/mingw, WSL, WSL2?
- From my understanding the `tls12-download.exe` should only be run in a `native windows environment` to download a signed vcpkg.exe for windows?
- On *nix environments, the `bootstrap-vcpkg.sh` downloads and builds the source from scratch as always has in the past.

**Comment by yuri-sevatz** [comment]
I am not running `bootstrap-vcpkg.bat` with Cygwin.  I am just used Cygwin's git to clone `microsoft/vcpkg` to my local hard drive, and then Windows' native shell to run `bootstrap-vcpkg.bat`.

This issue isn't so much whether `vcpkg` supports Cygwin, as it is whether `microsoft/vcpkg` is merging pre-compiled binaries that are meant to be executed, into git.  Git has file modes assigned during merges () and the commit I linked does not set execute permissions on what seems to be the only executable checked into `microsoft/vcpkg`.

```
$ git checkout 2020.11-1
HEAD is now at ec6fe06e8 Update CHANGELOG 2020.11.03 (#14377)
$ find ./ -name "*.exe"
(... nothing ...)
```
vs
```
$ git checkout master
Previous HEAD position was ec6fe06e8 Update CHANGELOG 2020.11.03 (#14377)
Switched to branch 'master'
Your branch is up to date with 'origin/master'.
$ find ./ -name "*.exe"
./scripts/tls12-download.exe
$
```

My guess is that whoever authored the commit in question may have been using a git client that was just misconfigured to drop the execute bit when merging.

**Comment by StarGate-One** [comment]
I am pretty sure Cygwin environment should not be executing the tls12-download.exe file - it is for a native windows only (run from cmd or powershell prompt only), which does not care about POSIX file permissions. Those permissions only apply to *ix environments, and last I checked *ix environments are not supposed to execute *.exe files.
Maybe @BillyONeal  and/or @ras0219-msft can shed some light on this?

**Comment by StarGate-One** [comment]
Note: All you need to do is go to this URL (https://github.com/microsoft/vcpkg-tool/releases) and download vcpkg.exe for windows. 
The sole purpose of tls12-download.exe is to download a signed version vcpkg.exe for windows and put in the [vcpkg-root].

**Comment by yuri-sevatz** [comment]
I was following the steps here:
https://docs.microsoft.com/en-us/cpp/build/install-vcpkg?view=msvc-160&tabs=windows

That aside, if I download every non-source assert available from this url: https://github.com/microsoft/vcpkg-tool/releases/ , and I deposit them into a brand new folder, I get:

```
C:\src\vcpkg-new>vcpkg.exe install termcolor:x64-windows
Error: Could not detect vcpkg-root
```

Judging from the looks of that, I don't think the new way of hosting `vcpkg` is quite ready yet, because there seems to be no information on how to install releases downloaded from it, and nobody seems to know how to use the new repository for standalone installations.  (I even had a spammer asking about what that repository was for in the very Issue I had filed for this PR :1st_place_medal: )

Maybe we could just fix the file mode checked into git, at least until the new repository is supported.  That way at least it would keep everything run smoothly under the old happy path, until you've got the new one up and running with all of Microsoft's lovely advertising.

**Comment by yuri-sevatz** [comment]
PS> I'm not complaining about the advertising ;)  I'm pretty sure you guys have won rewards for awards for developer documentation in the past!

**Comment by MVoz** [comment]
@yuri-sevatz
>  (I even had a spammer asking about what that repository was for in the very Issue I had filed for this PR 🥇 )
open CMD or POWERSHELL terminal, no problem. spammer )))
I don't think you know what you want, and that's a fact, according to your comments.

**Comment by yuri-sevatz** [comment]
@MVoz WHY ARE YOU FOLLOWING ME.

You're not even talking about the PR. 🤦‍♂️🤦‍♂️🤦‍♂️

**Comment by StarGate-One** [comment]
The documentation is a little vague in this area as I believe it is still being refined.
- From what feedback I have got back and figured out:
1. The windows bootstrap-vcpkg.bat runs the tls12-download.exe to download the latest release version of vcpkg.exe from the microsoft/vcpkg-tool repo
2. The linux/OSX bootstrap-vcpkg.sh clones the vcpkg-tool source to [vcpkg-root]/ports/_vcpkg directory then builds/compiles/links to create the vcpkg binary in the [vcpkg-root]
3. If you still want to build your own vcpkg.exe for windows, just clone the vcpkg-tool repo and use CMake to build the vcpkg.exe and copy it to the [vcpkg-root].
- For windows I choose the latter option 3, using CMake with -G Visual Studio 16 2019 -A x64 -T v142
- Bottom line, the binaries vcpkg.exe (windows) or vcpkg (everything else) is being built from the vcpkg-tool repo.

**Comment by yuri-sevatz** [comment]
@StarGate-One That's great information and it's all fine.

But I am just trying to make the fix I made for following Microsoft's steps on their website work consistently, for the benefit myself and others, so that nobody else has to go through any extra steps or understanding when cloning.

The permission change works fine, and it would be good to merge it for the reasons I described earlier.

**Comment by StarGate-One** [comment]
@yuri-sevatz I do not make these decisions. I make comments, help others understand where possible, submit changes in my free time. 
Ultimately it will be up to the MS project owners/sponsors as to whether this will be merged in, a change is required, etc.

**Review by BillyONeal — APPROVED** [review]

**Comment by BillyONeal** [comment]
> 2\. The linux/OSX bootstrap-vcpkg.sh clones the vcpkg-tool source to [vcpkg-root]/ports/_vcpkg directory then builds/compiles/links to create the vcpkg binary in the [vcpkg-root]

There's also a thing where we detect that we're in cygwin and dispatch to the .bat instead probably causing the confusion here https://github.com/microsoft/vcpkg/blob/5f29a7a5cae5374e99a488f3b324f1948de43c89/scripts/bootstrap.sh#L39

**Comment by yuri-sevatz** [comment]
> There's also a thing where we detect that we're in cygwin and dispatch to the .bat instead probably causing the confusion here

Well that's great news!  😄 I didn't want to push my luck since you never know with cygwin, but this actually works mostly well...

```
yuri-sevatz@satan ~/projects
$ git clone https://github.com/microsoft/vcpkg
Cloning into 'vcpkg'...
remote: Enumerating objects: 103779, done.
remote: Total 103779 (delta 0), reused 0 (delta 0), pack-reused 103779
Receiving objects: 100% (103779/103779), 31.45 MiB | 15.69 MiB/s, done.
Resolving deltas: 100% (64999/64999), done.
Updating files: 100% (7228/7228), done.

yuri-sevatz@satan ~/projects
$ cd vcpkg/

yuri-sevatz@satan ~/projects/vcpkg
$ ./bootstrap-vcpkg.bat
bash: ./bootstrap-vcpkg.bat: Permission denied

yuri-sevatz@satan ~/projects/vcpkg
$ chmod +x bootstrap-vcpkg.bat

yuri-sevatz@satan ~/projects/vcpkg
$ ./bootstrap-vcpkg.bat
Program 'tls12-download.exe' failed to run: Access is deniedAt C:\cygwin64\home\Jet\projects\vcpkg\scripts\bootstrap.ps1:48 char:1
+ & "$scriptsDir/tls12-download.exe" github.com "/microsoft/vcpkg-tool/ ...
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~.
At C:\cygwin64\home\Jet\projects\vcpkg\scripts\bootstrap.ps1:48 char:1
+ & "$scriptsDir/tls12-download.exe" github.com "/microsoft/vcpkg-tool/ ...
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : ResourceUnavailable: (:) [], ApplicationFailedException
    + FullyQualifiedErrorId : NativeCommandFailed


The variable '$LASTEXITCODE' cannot be retrieved because it has not been set.
At C:\cygwin64\home\Jet\projects\vcpkg\scripts\bootstrap.ps1:51 char:5
+ if ($LASTEXITCODE -ne 0)
+     ~~~~~~~~~~~~~
    + CategoryInfo          : InvalidOperation: (LASTEXITCODE:String) [], RuntimeException
    + FullyQualifiedErrorId : VariableIsUndefined

Telemetry
---------
vcpkg collects usage data in order to help us improve your experience.
The data collected by Microsoft is anonymous.
You can opt-out of telemetry by re-running the bootstrap-vcpkg script with -disableMetrics,
passing --disable-metrics to vcpkg on the command line,
or by setting the VCPKG_DISABLE_METRICS environment variable.

Read more about vcpkg telemetry at docs/about/privacy.md

```

^^ Similar issue  with no execute bit.   Pushing past this....

```
yuri-sevatz@satan ~/projects/vcpkg
$ git remote add yuri-sevatz git@github.com:yuri-sevatz/vcpkg.git

yuri-sevatz@satan ~/projects/vcpkg
$ git fetch yuri-sevatz
remote: Enumerating objects: 4, done.
remote: Counting objects: 100% (4/4), done.
remote: Total 6 (delta 4), reused 4 (delta 4), pack-reused 2
Unpacking objects: 100% (6/6), 675 bytes | 24.00 KiB/s, done.
From github.com:yuri-sevatz/vcpkg
 * [new branch]          develop    -> yuri-sevatz/develop
 * [new branch]          master     -> yuri-sevatz/master

yuri-sevatz@satan ~/projects/vcpkg
$ git checkout -b develop yuri-sevatz/develop
M       bootstrap-vcpkg.bat
Branch 'develop' set up to track remote branch 'develop' from 'yuri-sevatz'.
Switched to a new branch 'develop'

yuri-sevatz@satan ~/projects/vcpkg
$ ./bootstrap-vcpkg.bat
Downloading https://github.com/microsoft/vcpkg-tool/releases/download/2021-01-13-768d8f95c9e752603d2c5901c7a7c7fbdb08af35/vcpkg.exe -> C:\cygwin64\home\Jet\projects\vcpkg/vcpkg.exe
Done.

Telemetry
---------
vcpkg collects usage data in order to help us improve your experience.
The data collected by Microsoft is anonymous.
You can opt-out of telemetry by re-running the bootstrap-vcpkg script with -disableMetrics,
passing --disable-metrics to vcpkg on the command line,
or by setting the VCPKG_DISABLE_METRICS environment variable.

Read more about vcpkg telemetry at docs/about/privacy.md

```

Success!

```
yuri-sevatz@satan ~/projects/vcpkg
$ ./vcpkg.exe install termcolor:x64-windows
Computing installation plan...
The following packages will be built and installed:
    termcolor[core]:x64-windows -> 1.0.1
Detecting compiler hash for triplet x64-windows...
A suitable version of powershell-core was not found (required v7.1.0). Downloading portable powershell-core v7.1.0...
Downloading powershell-core...
  https://github.com/PowerShell/PowerShell/releases/download/v7.1.0/PowerShell-7.1.0-win-x86.zip -> C:\cygwin64\home\Jet\projects\vcpkg\downloads\PowerShell-7.1.0-win-x86.zip
Extracting powershell-core...
A suitable version of 7zip was not found (required v18.1.0). Downloading portable 7zip v18.1.0...
Downloading 7zip...
  https://www.nuget.org/api/v2/package/7-Zip.CommandLine/18.1.0 -> C:\cygwin64\home\Jet\projects\vcpkg\downloads\7-zip.commandline.18.1.0.nupkg
Extracting 7zip...
A suitable version of nuget was not found (required v5.5.1). Downloading portable nuget v5.5.1...
Downloading nuget...
  https://dist.nuget.org/win-x86-commandline/v5.5.1/nuget.exe -> C:\cygwin64\home\Jet\projects\vcpkg\downloads\22ea847d-nuget.exe
Using cached binary package: C:\Users\Jet\AppData\Local\vcpkg\archives\1b\1b9e13a0297a4df84330cf64f98791c9d751805e.zip
Starting package 1/1: termcolor:x64-windows
Building package termcolor[core]:x64-windows...
Building package termcolor[core]:x64-windows... done
Installing package termcolor[core]:x64-windows...
Installing package termcolor[core]:x64-windows... done
Elapsed time for package termcolor:x64-windows: 107.6 ms

Total elapsed time: 18.69 s

The package termcolor:x64-windows provides CMake targets:

    find_package(termcolor CONFIG REQUIRED)
    target_link_libraries(main PRIVATE termcolor::termcolor)
```

^^ Everything else working as expected.

```
yuri-sevatz@satan ~/projects/vcpkg
$ git diff
diff --git a/bootstrap-vcpkg.bat b/bootstrap-vcpkg.bat
old mode 100644
new mode 100755
```

We can add a file mode 100644 -> 100755 for `./bootstrap-vcpkg.bat` to this branch too if @BillyONeal wants ?

**Comment by ras0219-msft** [comment]
Thanks for the fix!

If changing the execute bit on `bootstrap-vcpkg.bat` makes things work correctly in cygwin, that'd be great and I'd appreciate another PR :)
